src/include/self_driving/model_server/model_server_manager.h

```diff
@@ -0,0 +1,239 @@
+  Result result_;
+
+  /** Condition variable **/
+  std::condition_variable cvar_;
+
+  /** If async operation done */
+  std::atomic<bool> done_ = false;
+
+  /** If async operation succeeds */
+  std::atomic<bool> success_ = false;
+
+  /** Reason for failure **/
+  std::string fail_msg_;
+
+  /** Mutex associated with the condition variable **/
+  std::mutex mtx_;
+};
+
+/**
+ * Interface for ModelServerManager related operations
+ */
+class ModelServerManager {
+  enum class Callback : uint64_t { NOOP = 0, CONNECTED, DEFAULT };
+
+ public:
+  /**
+   * Construct a ModelServerManager with the given executable script to the python ModelServer
+   * @param model_bin python script path
+   * @param messenger Messenger pointer
+   */
+  ModelServerManager(const std::string &model_bin, const common::ManagedPointer<messenger::Messenger> &messenger);
+
+  /**
+   * Stop the python ModelServer when exits
+   */
+  ~ModelServerManager() { StopModelServer(); }
+
+  /**
+   * Stop the python-ModelServer daemon by sending a message to the Python model server
+   */
+  void StopModelServer();
+
+  /**
+   * Check if the model server has started.
+   * The user of this function should poll this until the pid changes.
+   *
+   * A true return value means the model server script is running and ready to receive message.
+   *
+   * @return true if model server has started
+   */
+  bool ModelServerStarted() const { return connected_; }
+
+  /**
+   * Get the python model-server's PID
+   * @return  pid
+   */
+  pid_t GetModelPid() const { return py_pid_; }
+
+  /*******************************************************
+   * ModelServer <-> ModelServerManager logic routines
+   *******************************************************/
+
+  /**
+   * Ask the model server to print a message
+   * This function does not expect a callback
+   *
+   * @warning The message should not contain '-' as per current protocol
+   * @param msg Message to print
+   */
+  void PrintMessage(const std::string &msg);
+
+  /**
+   * Train a model with the given seq files directory
+   *
+   *  This function will be invoked asynchronously.
+   *  The caller should poll from ModelServerManager::Trained() to check if the training has been done.
+   *
+   *  The caller should use the save_path as a handle to the trained model for inference later on.
+   *
+   * @param methods list of candidates methods that will be used for training
+   * @param save_path path to where the trained model map will be stored at
+   * @param seq_files_dir Seq files's enclosing directory (seq files generated by mini_runner)
+   * @param future A future object which the caller waits for
+   */
+  void TrainWith(const std::vector<std::string> &methods, const std::string &seq_files_dir,
+                 const std::string &save_path, ModelServerFuture<std::string> *future);
+
+  /**
+   * Perform inference on the given data file.
+   *
+   * This function should expect some return (Not figured out yet)
```

Comment:
Sorry, can you document this a bit more? What is currently being returned and what is being figured out?

Reply:
Also, this function is blocking, right? Can you add a quick note to all blocking functions' header comments, that they block until the result is available?